RiscVPkg/Universal/OpenSbi/Pei/OpenSbiPeim.c

```diff
@@ -0,0 +1,499 @@
+  0x00,
+  PeiOpenSbiVendorCall
+};
+
+EFI_PEI_PPI_DESCRIPTOR mOpenSbiVendorPpiDesc = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gPeiOpenSbiVendorPpiGuid,
+  &mOpenSbiVendorPpi
+};
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiLibraryScratchThisHartPtr (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  OUT struct sbi_scratch            **ScratchPtr
+  )
+{
+    *ScratchPtr = (struct sbi_scratch *)csr_read(CSR_MSCRATCH);
+    // TODO: Can this ever fail?
+    return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiLibraryPlatformPtr (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  IN  struct sbi_scratch             *ScratchPtr,
+  OUT const struct sbi_platform     **PlatformPtr
+  )
+{
+    if (ScratchPtr == NULL)
+        return EFI_INVALID_PARAMETER;
+
+    *PlatformPtr = (const struct sbi_platform *)(ScratchPtr->platform_addr);
+    // TODO: Do we need this function at all? In OpenSBI it's just a macro.
+    return EFI_SUCCESS;
+}
+
+typedef struct sbi_scratch *(*h2s)(ulong hartid);
+
+EFI_STATUS
+EFIAPI
+PeiOpenSbiLibraryHartIdToScratch (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  IN  struct sbi_scratch             *AnyScratchPtr,
+  IN  UINTN                           HartId,
+  OUT const struct sbi_scratch      **ScratchPtr
+  )
+{
+    if (AnyScratchPtr == NULL)
+        return EFI_INVALID_PARAMETER;
+
+    *ScratchPtr = ((h2s)AnyScratchPtr->hartid_to_scratch)(HartId);
+
+    // TODO: Maybe check if *ScratchPtr is NULL?
+    return EFI_SUCCESS;
+}
+
+VOID
+EFIAPI
+PeiOpenSbiLibrarySbiInit (
+  IN  EFI_PEI_SERVICES              **PeiServices,
+  IN  PEI_RISCV_OPENSBI_LIBRARY_PPI  *This,
+  IN  struct sbi_scratch             *ScratchPtr
+  )
+{
+    sbi_init(ScratchPtr);
+    // TOOD: Never returns. Should we put an annotation?
+}
+
+//
+// OpenSBI PPI interface function for the OpenSBI library itself
+// Can only be called from M-Mode!
+//
+PEI_RISCV_OPENSBI_LIBRARY_PPI mOpenSbiLibraryPpi = {
+  0x00,
+  PeiOpenSbiLibraryScratchThisHartPtr,
+  PeiOpenSbiLibraryPlatformPtr,
+  PeiOpenSbiLibraryHartIdToScratch,
+  PeiOpenSbiLibrarySbiInit
+};
+
+EFI_PEI_PPI_DESCRIPTOR mOpenSbiLibraryPpiDesc = {
+  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
+  &gPeiOpenSbiLibraryPpiGuid,
+  &mOpenSbiLibraryPpi
+};
+
+/**
+  Entry point of this module.
+
+  @param[in] FileHandle   Handle of the file being invoked.
+  @param[in] PeiServices  Describes the list of possible PEI Services.
+
+  @return Status.
+
+**/
+EFI_STATUS
+EFIAPI
+OpenSbiPeimEntry (
+  IN       EFI_PEI_FILE_HANDLE   FileHandle,
+  IN CONST EFI_PEI_SERVICES    **PeiServices
+  )
+{
+  EFI_STATUS                        Status;
+
+  // TODO: Check status of each
+  // TODO: Abner: You can install these PPIs at once, with EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST set in the last entry.
```

Comment:
cc @changab How?

Reply:
EFI_PEI_PPI_DESCRIPTOR mABCPpi []= {
 {
  (EFI_PEI_PPI_DESCRIPTOR_PPI ,
  &g_A_PpiGuid,
  (VOID *)&m_A_Ppi
 },
  (EFI_PEI_PPI_DESCRIPTOR_PPI ,
  &g_B_PpiGuid,
  (VOID *)&m_B_Ppi
 },
  (EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST),
  &g_C_PpiGuid,
  (VOID *)&m_C_Ppi
 }
};

Reply:
Oh! I looked at many PPIs and none of them did it this way, I'll do that.

Reply:
Done